<a href="https://colab.research.google.com/github/akshatchaturvedi28/Collab-TensorFlow/blob/master/CNNs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CONVOLUTIONAL NEURAL NETWORKS (CNNs)**

In [0]:
!pip install tensorflow==2.0.0-beta1
import tensorflow as tf

In [0]:
from __future__ import absolute_import, division, print_function, unicode_literals
from tensorflow.keras import datasets, layers, models

import datetime, os


In [0]:
(train_images, train_labels), (test_images, test_labels) = datasets.mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


Reshaping 3D tensors to 4D tensors to satisfy CNN requirements

In [0]:
train_images = train_images.reshape((60000,28,28,1))
test_images = test_images.reshape((10000,28,28,1))

**Create a convolutional base**

In [0]:
models = models.Sequential()

As input a CNN takes the tensors of a shape (image_height, image_width, color_channels), ignoring the batch size. MNIST (has 1 color_channels) because images are in grayscale. In colored images we have 3 channels (red, green, blue)

In [0]:
models.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape = (28,28,1)))

In [0]:
models.add(layers.MaxPooling2D((2, 2)))

In [0]:
models.add(layers.Conv2D(64, (3, 3), activation='relu'))

In [0]:
models.add(layers.MaxPooling2D((2, 2)))

In [0]:
models.add(layers.Conv2D(64, (3, 3), activation='relu'))

In [0]:
models.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


In [0]:
models.add(layers.Flatten())
models.add(layers.Dense(64, activation='relu'))
models.add(layers.Dense(10, activation='softmax'))

In [0]:
models.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten (Flatten)            (None, 576)               0         
_________________________________________________________________
dense (Dense)                (None, 64)                3

In [0]:
models.compile(optimizer='adam',
               loss='sparse_categorical_crossentropy',
               metrics=['accuracy'])
models.fit(train_images, train_labels, epochs=5)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 43s 725us/sample - loss: 0.3223 - accuracy: 0.9453
Epoch 2/5
60000/60000 [==============================] - 43s 713us/sample - loss: 0.0619 - accuracy: 0.9817
Epoch 3/5
60000/60000 [==============================] - 44s 731us/sample - loss: 0.0463 - accuracy: 0.9859
Epoch 4/5
60000/60000 [==============================] - 44s 738us/sample - loss: 0.0398 - accuracy: 0.9877
Epoch 5/5
60000/60000 [==============================] - 44s 734us/sample - loss: 0.0337 - accuracy: 0.9902


**EVALUATE THE MODEL**

In [0]:
test_loss, test_acc = models.evaluate(test_images, test_labels)
print(test_acc)

10000/10000 [==============================] - 2s 240us/sample - loss: 0.0447 - accuracy: 0.9888
0.9888


Our simple CNN has achieved an accuracy of 99% approx.

## **Classification with Feed Forward Networks**

In [0]:
from tensorflow import keras
from keras import layers

In [0]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train/255.0, x_test/255.0

In [0]:
model_fnn = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation='softmax')
])

In [0]:
model_fnn.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [0]:
model_fnn.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


In [0]:
model_fnn.fit(x_train, y_train, epochs=5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 4s 66us/sample - loss: 0.2948 - accuracy: 0.9134
Epoch 2/5
60000/60000 [==============================] - 4s 69us/sample - loss: 0.1402 - accuracy: 0.9578
Epoch 3/5
60000/60000 [==============================] - 4s 67us/sample - loss: 0.1070 - accuracy: 0.9673
Epoch 4/5
60000/60000 [==============================] - 4s 64us/sample - loss: 0.0865 - accuracy: 0.9733
Epoch 5/5
60000/60000 [==============================] - 4s 66us/sample - loss: 0.0755 - accuracy: 0.9771


In [0]:
test_acc1, test_loss1 = model_fnn.evaluate(x_test, y_test)
print(test_acc1)

10000/10000 [==============================] - 0s 38us/sample - loss: 0.0716 - accuracy: 0.9777
0.07157029427578673


Our Accuracy in Feed Forward NN is 97% approx.